### PR TITLE
Add a generator for Karaf based projects #178

### PR DESCRIPTION
**Pull request by lburgazzoli**

### CODE REVIEW

**Comment by rhuss** [comment]
Thanks ;-) will merge it today and perform a new release ...

File: generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        combine        {{ d = "false"; }},
+        name           {{ d = "%g/%a:%l"; }},
+        alias          {{ d = "karaf"; }},
+        baseImage      {{ d = "fabric8/s2i-karaf:1.3"; }},
+        baseDir        {{ d = "/deployments/"; }},
+        user           {{ d = "jboss:jboss:jboss"; }},
+        cmd            {{ d = "/deployments/deploy-and-run.sh"; }},
+        webPort        {{ d = "8181"; }},
+        jolokiaPort    {{ d = "8778"; }},
+        prometheusPort {{ d = "9779"; }};
```
Comment (rhuss) [inline generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java]: AFAIK the s2i-karaf base images does not use agent-bond (and hence does not export Prometheus' jmx_exporter). Guess we can savely remove this here.

File: generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        public String def() { return d; } protected String d;
+    }
+
+    @Override
+    public List<ImageConfiguration> customize(List<ImageConfiguration> configs) {
```
Comment (rhuss) [inline generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java]: The general structure looks quite similar to the Spring Boot generator. I wonder where we should extract common functionality into a base class ? The question then would be where to put this class. 

Or maybe its not worth the effort ? I see at least things like isApplicable() or shouldIncludeDefaultImage() which could be helpful in the `BaseGenerator`.

File: generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java (hunk 1 above)
Comment (lburgazzoli) [inline generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java]: oh yes. looks like I kept it from spring boot one

File: generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import io.fabric8.maven.core.util.MavenUtil;
+import io.fabric8.maven.docker.config.AssemblyConfiguration;
+import io.fabric8.maven.docker.config.BuildImageConfiguration;
+import io.fabric8.maven.docker.config.ImageConfiguration;
+import io.fabric8.maven.generator.api.BaseGenerator;
+import io.fabric8.maven.generator.api.MavenGeneratorContext;
+import io.fabric8.utils.Strings;
+import org.apache.maven.project.MavenProject;
+
+public class KarafGenerator extends BaseGenerator {
```
Comment (rhuss) [inline generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java]: I know, we don't have yet a good unittest coverage, but I push hard to improve that. Currently we are around 12% (see also https://sonarqube.com/overview?id=io.fabric8%3Afabric8-maven-plugin-build). 

I would be **awesome** if you could provide a unit test for the generator. If you need a mocking framework we use [JMockit](http://jmockit.org/).

File: generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java (hunk 2 above)
Comment (lburgazzoli) [inline generator/karaf/src/main/java/io/fabric8/maven/generator/karaf/KarafGenerator.java]: I think it makes sense, i.e. the isApplicable could be used also by the manager so we can log is a generator has been applied or not

**Comment by rhuss** [comment]
Look good to me (except for some minor things, see comments). 

If you dont have much time, we could merge it now and I could do the refactoring since I'm going to have todo some refactorings soon anyway. But I'm also very happy if you could address them ;-)

**Comment by lburgazzoli** [comment]
Do you mean also move isApplicable and shouldIncludeDefaultImage to BaseGenerator ?
I'm not sure about providing tests by today but I can definitively work on the other things

**Comment by rhuss** [comment]
would be fine for me, we can add the unit test later (which is what we always say ;-)

I would the merge it into master.

**Comment by lburgazzoli** [comment]
@rhuss can you have a look at the latest code ?

**Comment by rhuss** [comment]
Sorry, had to leave early. Will review tomorrow

...roland

Luca Burgazzoli <notifications@github.com> schrieb am Mi., 27. Juli 2016,
13:04:

**Comment by rhuss** [comment]
LGTM

Thanks a lot ! I might going to adapt it later a bit because of unifying the generator stuff (e.g. possibly moving the 'combine' feature up to the base generator.

**Comment by lburgazzoli** [comment]
thx

**Pull request merged**